core/src/apps/common/sdcard.py

```diff
@@ -45,17 +45,57 @@ class SdProtectCancelled(wire.ProcessError):
     return await confirm(ctx, text, confirm=btn_confirm, cancel=btn_cancel)
 
 
+async def format_card_dialog(ctx: wire.GenericContext) -> bool:
+    # Format card? yes/no
+    text = Text("SD card error", ui.ICON_WRONG, ui.RED)
+    text.bold("Unknown filesystem.")
+    text.br_half()
+    text.normal("Use a different card or")
+    text.normal("format the SD card to")
+    text.normal("the FAT32 filesystem.")
+    if not await confirm(ctx, text, confirm="Format", cancel="Cancel"):
+        return False
+
+    # Confirm formatting
+    text = Text("Format SD card", ui.ICON_WIPE, ui.RED)
+    text.normal("Do you really want to", "format the SD card?")
+    text.br_half()
+    text.bold("All data on the SD card", "will be lost.")
+    return await hold_to_confirm(ctx, text, confirm="Format SD card")
+
+
 async def sd_problem_dialog(ctx: wire.GenericContext) -> bool:
-    text = Text("SD card protection", ui.ICON_WRONG, ui.RED)
+    text = Text("SD card problem", ui.ICON_WRONG, ui.RED)
     text.normal("There was a problem", "accessing the SD card.")
     return await confirm(ctx, text, confirm="Retry", cancel="Abort")
 
 
-async def ensure_sd_card(ctx: wire.GenericContext) -> None:
-    sd = io.SDCard()
-    while not sd.present():
+async def ensure_sdcard(ctx: wire.GenericContext) -> None:
+    while not sdcard.is_present():
         if not await insert_card_dialog(ctx):
-            raise SdProtectCancelled("SD card required.")
+            raise SdCardUnavailable("SD card required.")
+
+    while True:
+        try:
+            with sdcard.get_filesystem(mounted=False) as fs:
+                fs.mount()
+                # Mount succeeded, filesystem is OK
+                return
+        except OSError:
+            # Mount failed. Handle the problem outside except-clause
+            pass
+
+        if not await format_card_dialog(ctx):
+            raise SdCardUnavailable("SD card not formatted.")
+
+        try:
+            with sdcard.get_filesystem(mounted=False) as fs:
+                fs.mkfs()
+                # mkfs succeeded. Re-run loop to retry mounting.
+                continue
+        except OSError:
+            if not await sd_problem_dialog(ctx):
+                raise SdCardUnavailable("Problem formatting SD card.")
```

Comment:
Nit-again: It seems weird to me we are suddenly using the word "problem" instead of "error".

Reply:
i think it's appropriate in "sd_problem_dialog", because it indicates a generic "problem" which is not necessarily an error.

no preference re the user-facing string, we can safely switch to "error" there